kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -79,25 +82,38 @@ internal class KotlinJsonAdapter<T>(
             "Multiple values for '${binding.property.name}' at ${reader.path}")
       }
 
-      values[index] = binding.adapter.fromJson(reader)
+      values[index] = when (failureStrategy) {
+        FailureStrategy.FAIL_FAST -> binding.adapter.fromJson(reader)
+        FailureStrategy.ACCUMULATE_ERRORS -> binding.adapter.nullSafe().fromJson(reader)
```

Comment:
`nullSafe()` because otherwise it throws `JsonDataException` on the first null

Reply:
This is going to allocate every time it's called, which is going to be a lot. This needs to be done at binding creation.

Reply:
is `nullSafe()` significantly worse than the other version? This was always created here, I just added the `when`.

Reply:
What other version? The original version just looks up the adapter instance on a field.

Reply:
Old code:
`values[index] = binding.adapter.fromJson(reader)`

new code:

```
values[index] = when (failureStrategy) {
        FailureStrategy.FAIL_FAST -> binding.adapter.fromJson(reader)
        FailureStrategy.ACCUMULATE_ERRORS -> binding.adapter.nullSafe().fromJson(reader)
}
```

is `binding.adapter.nullSafe().fromJson(reader)` significantly worse than `binding.adapter.fromJson(reader)` ?

> This is going to allocate every time it's called, which is going to be a lot.

is that the allocation you are referring to?

Reply:
Yes